File: ParamFileGenerator/Parameters/Params.vb
Imports System.IO
Imports System.Linq

Public Class Params
    Implements IBasicParams
    Implements IAdvancedParams

    ' Ignore Spelling: Xcalibur

    Public Enum ParamFileTypes
        BioWorks_20 = 0 'Normal BioWorks 2.0 SEQUEST
        BioWorks_30 = 1 'BioWorks 3.0+ TurboSEQUEST
        BioWorks_31 = 2 'BioWorks 3.1 ClusterQuest
        BioWorks_32 = 3 'BioWorks 3.2 ClusterF***
    End Enum

    Const DEF_DB_NAME As String = "C:\Xcalibur\database\nr.fasta"     'Not really used, just a placeholder

    Const DEF_ENZ_SECTION_NAME As String = "SEQUEST_ENZYME_INFO"

    'Basic Parameters
    Private m_type As ParamFileTypes

    'Advanced Parameters
    Private m_ionSeriesString As String
    Private m_protMassFilterString As String

    Private m_fullTemplate As RetrieveParams
    Private m_templateFilePath As String

    Public Shared ReadOnly Property BaseLineParamSet As Params
        Get
            Return MainProcess.BaseLineParamSet
        End Get
    End Property

    Public ReadOnly Property FileType As ParamFileTypes Implements IBasicParams.FileType
        Get
            Return m_type
        End Get
    End Property

    Public Property DMS_ID As Integer Implements IBasicParams.DMS_ID

    Public Property FileName As String Implements IBasicParams.FileName

    Public Property FileTypeIndex As Integer

    Public Property Description As String Implements IBasicParams.Description

    Public Property DefaultFASTAPath As String Implements IAdvancedParams.DefaultFASTAPath

    Public Property DefaultFASTAPath2 As String Implements IAdvancedParams.DefaultFASTAPath2

    Public Property NumberOfResultsToProcess As Integer Implements IAdvancedParams.NumberOfResultsToProcess

    Public Property PeptideMassTolerance As Single Implements IAdvancedParams.PeptideMassTolerance

    Public Property CreateOutputFiles As Boolean Implements IAdvancedParams.CreateOutputFiles

    Public Property IonSeries As IonSeries Implements IAdvancedParams.IonSeries

    Public Property DynamicMods As DynamicMods Implements IBasicParams.DynamicMods

    Public Property TermDynamicMods As TermDynamicMods Implements IBasicParams.TermDynamicMods

    Public Property IsotopicMods As IsoMods Implements IBasicParams.IsotopicModificationsList

    Public Property MaximumNumAAPerDynMod As Integer Implements IAdvancedParams.MaximumNumAAPerDynMod

    Public Property MaximumNumDifferentialPerPeptide As Integer Implements IAdvancedParams.MaximumDifferentialPerPeptide

    Public Property UsePhosphoFragmentation As Integer Implements IAdvancedParams.UsePhosphoFragmentation

    Public Property FragmentIonTolerance As Single Implements IAdvancedParams.FragmentIonTolerance

    Public Property NumberOfOutputLines As Integer Implements IAdvancedParams.NumberOfOutputLines

    Public Property NumberOfDescriptionLines As Integer Implements IAdvancedParams.NumberOfDescriptionLines

    Public Property ShowFragmentIons As Boolean Implements IAdvancedParams.ShowFragmentIons

    Public Property PrintDuplicateReferences As Boolean Implements IAdvancedParams.PrintDuplicateReferences

    Public Property SelectedEnzymeDetails As EnzymeDetails Implements IBasicParams.SelectedEnzymeDetails

    Public Property SelectedEnzymeIndex As Integer Implements IBasicParams.SelectedEnzymeIndex

    Public Property SelectedEnzymeCleavagePosition As Integer Implements IBasicParams.SelectedEnzymeCleavagePosition

    Public Property SelectedNucReadingFrame As IAdvancedParams.FrameList Implements IAdvancedParams.SelectedNucReadingFrame

    Public Property SelectedNucReadingFrameIndex As Integer Implements IAdvancedParams.SelectedNucReadingFrameIndex
        Get
            Return SelectedNucReadingFrame
        End Get
        Set
            SelectedNucReadingFrame = CType(Value, IAdvancedParams.FrameList)
        End Set
    End Property

    Public Property ParentMassType As IBasicParams.MassTypeList Implements IBasicParams.ParentMassType

    Public Property FragmentMassType As IBasicParams.MassTypeList Implements IBasicParams.FragmentMassType

    Public Property RemovePrecursorPeak As Boolean Implements IAdvancedParams.RemovePrecursorPeak

    Public Property IonCutoffPercentage As Single Implements IAdvancedParams.IonCutoffPercentage

    Public Property MaximumNumberMissedCleavages As Integer Implements IBasicParams.MaximumNumberMissedCleavages

    Public Property MinimumProteinMassToSearch As Single Implements IAdvancedParams.MinimumProteinMassToSearch

    Public Property MaximumProteinMassToSearch As Single Implements IAdvancedParams.MaximumProteinMassToSearch

    Public Property NumberOfDetectedPeaksToMatch As Integer Implements IAdvancedParams.NumberOfDetectedPeaksToMatch

    Public Property NumberOfAllowedDetectedPeakErrors As Integer Implements IAdvancedParams.NumberOfAllowedDetectedPeakErrors

    Public Property MatchedPeakMassTolerance As Single Implements IAdvancedParams.MatchedPeakMassTolerance

    Public Property AminoAcidsAllUpperCase As Boolean Implements IAdvancedParams.AminoAcidsAllUpperCase

    Public Property PartialSequenceToMatch As String Implements IBasicParams.PartialSequenceToMatch

    Public Property SequenceHeaderInfoToFilter As String Implements IAdvancedParams.SequenceHeaderInfoToFilter

    Public Property PeptideMassUnits As Integer Implements IAdvancedParams.PeptideMassUnits

    Public Property FragmentMassUnits As Integer Implements IAdvancedParams.FragmentMassUnits

    Public Property StaticModificationsList As StaticMods Implements IBasicParams.StaticModificationsList

    Public Property EnzymeList As EnzymeCollection Implements IBasicParams.EnzymeList

    Public Property LoadedParamNames As Hashtable = New Hashtable

    Public Sub AddLoadedParamName(ParameterName As String, ParameterValue As String)
        If LoadedParamNames Is Nothing Then
            LoadedParamNames = New Hashtable
        End If
        If Not LoadedParamNames.ContainsKey(ParameterName) Then
            LoadedParamNames.Add(ParameterName, ParameterValue)
        End If
    End Sub

    Public Function RetrieveEnzymeDetails(EnzymeListIndex As Integer) As EnzymeDetails Implements IBasicParams.RetrieveEnzymeDetails
        Return EnzymeList.Item(EnzymeListIndex)
    End Function

    ''' <summary>
    ''' Constructor
    ''' </summary>
    Public Sub New()
        IonSeries = New IonSeries()
        EnzymeList = New EnzymeCollection()
        SelectedEnzymeDetails = New EnzymeDetails()
        DynamicMods = New DynamicMods()
        StaticModificationsList = New StaticMods()
        IsotopicMods = New IsoMods()
        TermDynamicMods = New TermDynamicMods("0.0 0.0")

        MaximumNumDifferentialPerPeptide = 3
        UsePhosphoFragmentation = 0
        PeptideMassUnits = IAdvancedParams.MassUnitList.amu
    End Sub

    Public Function ReturnMassFilter(MinimumMassToFilter As Single, MaximumMassToFilter As Single) As String
        Return ReturnMassFilterString(MinimumMassToFilter, MaximumMassToFilter)
    End Function

    Public Sub LoadTemplate(templateFileName As String)
        LoadTemplateParams(templateFileName)
    End Sub

    Private Sub LoadTemplateParams(templateFileName As String)

        m_templateFilePath = GetFilePath(templateFileName)
        Dim m_getEnzymeList As New GetEnzymeBlock(m_templateFilePath, DEF_ENZ_SECTION_NAME)
        EnzymeList = m_getEnzymeList.EnzymeList

        m_type = GetTemplateType()

        Dim SectionName = "SEQUEST"

        m_fullTemplate = New RetrieveParams(m_templateFilePath)
        'Retrieve Basic Parameters
        With m_fullTemplate
            .SetSection(SectionName)
            FileName = Path.GetFileName(m_templateFilePath)
            SelectedEnzymeIndex = CInt(.GetParam("enzyme_number"))
            SelectedEnzymeDetails = EnzymeList(SelectedEnzymeIndex)
            SelectedEnzymeCleavagePosition = 1
            MaximumNumberMissedCleavages = CInt(.GetParam("max_num_internal_cleavage_sites"))
            ParentMassType = CType(CInt(.GetParam("mass_type_parent")), IBasicParams.MassTypeList)
            FragmentMassType = CType(CInt(.GetParam("mass_type_fragment")), IBasicParams.MassTypeList)
            PartialSequenceToMatch = .GetParam("partial_sequence")
            DynamicMods = New DynamicMods(.GetParam("diff_search_options"))
            TermDynamicMods = New TermDynamicMods(.GetParam("term_diff_search_options"))
            StaticModificationsList = New StaticMods


            'Get Static Mods
            StaticModificationsList.CtermPeptide = CSng(.GetParam("add_Cterm_peptide"))
            StaticModificationsList.CtermProtein = CSng(.GetParam("add_Cterm_protein"))
            StaticModificationsList.NtermPeptide = CSng(.GetParam("add_Nterm_peptide"))
            StaticModificationsList.NtermProtein = CSng(.GetParam("add_Nterm_protein"))
            StaticModificationsList.G_Glycine = CSng(.GetParam("add_G_Glycine"))
            StaticModificationsList.A_Alanine = CSng(.GetParam("add_A_Alanine"))
            StaticModificationsList.S_Serine = CSng(.GetParam("add_S_Serine"))
            StaticModificationsList.P_Proline = CSng(.GetParam("add_P_Proline"))
            StaticModificationsList.V_Valine = CSng(.GetParam("add_V_Valine"))
            StaticModificationsList.T_Threonine = CSng(.GetParam("add_T_Threonine"))
            StaticModificationsList.C_Cysteine = CSng(.GetParam("add_C_Cysteine"))
            StaticModificationsList.L_Leucine = CSng(.GetParam("add_L_Leucine"))
            StaticModificationsList.I_Isoleucine = CSng(.GetParam("add_I_Isoleucine"))
            StaticModificationsList.X_LorI = CSng(.GetParam("add_X_LorI"))
            StaticModificationsList.N_Asparagine = CSng(.GetParam("add_N_Asparagine"))
            StaticModificationsList.O_Ornithine = CSng(.GetParam("add_O_Ornithine"))
            StaticModificationsList.B_avg_NandD = CSng(.GetParam("add_B_avg_NandD"))
            StaticModificationsList.D_Aspartic_Acid = CSng(.GetParam("add_D_Aspartic_Acid"))
            StaticModificationsList.Q_Glutamine = CSng(.GetParam("add_Q_Glutamine"))
            StaticModificationsList.K_Lysine = CSng(.GetParam("add_K_Lysine"))
            StaticModificationsList.Z_avg_QandE = CSng(.GetParam("add_Z_avg_QandE"))
            StaticModificationsList.E_Glutamic_Acid = CSng(.GetParam("add_E_Glutamic_Acid"))
            StaticModificationsList.M_Methionine = CSng(.GetParam("add_M_Methionine"))
            StaticModificationsList.H_Histidine = CSng(.GetParam("add_H_Histidine"))
            StaticModificationsList.F_Phenylalanine = CSng(.GetParam("add_F_Phenylalanine"))
            StaticModificationsList.R_Arginine = CSng(.GetParam("add_R_Arginine"))
            StaticModificationsList.Y_Tyrosine = CSng(.GetParam("add_Y_Tyrosine"))
            StaticModificationsList.W_Tryptophan = CSng(.GetParam("add_W_Tryptophan"))
        End With

        'add code to check for existence of isotopic mods



        'Retrieve Advanced Parameters
        With m_fullTemplate
            .SetSection(SectionName)
            If m_type = ParamFileTypes.BioWorks_20 Then
                DefaultFASTAPath = .GetParam("database_name")
                DefaultFASTAPath2 = ""
                NumberOfResultsToProcess = 500
            ElseIf m_type = ParamFileTypes.BioWorks_30 Then
                DefaultFASTAPath = .GetParam("first_database_name")
                DefaultFASTAPath2 = .GetParam("second_database_name")
                NumberOfResultsToProcess = CInt(.GetParam("num_results"))
            End If
            PeptideMassTolerance = CSng(.GetParam("peptide_mass_tolerance"))
            If .GetParam("create_output_files") IsNot Nothing Then
                CreateOutputFiles = CBool(.GetParam("create_output_files"))
            Else
                CreateOutputFiles = True
            End If
            m_ionSeriesString = .GetParam("ion_series")
            IonSeries = New IonSeries(m_ionSeriesString)
            MaximumNumAAPerDynMod = CInt(.GetParam("max_num_differential_AA_per_mod"))
            If m_type = ParamFileTypes.BioWorks_32 Then
                MaximumNumDifferentialPerPeptide = CInt(.GetParam("max_num_differential_per_peptide"))
            End If
            FragmentIonTolerance = CSng(.GetParam("fragment_ion_tolerance"))
            NumberOfOutputLines = CInt(.GetParam("num_output_lines"))
            NumberOfDescriptionLines = CInt(.GetParam("num_description_lines"))
            ShowFragmentIons = CBool(.GetParam("show_fragment_ions"))
            PrintDuplicateReferences = CBool(.GetParam("print_duplicate_references"))
            SelectedNucReadingFrame = CType(CInt(.GetParam("enzyme_number")), IAdvancedParams.FrameList)
            RemovePrecursorPeak = CBool(.GetParam("remove_precursor_peak"))
            IonCutoffPercentage = CSng(.GetParam("ion_cutoff_percentage"))
            m_protMassFilterString = .GetParam("protein_mass_filter")

            Dim protMassFilterList = m_protMassFilterString.Split(" "c)
            If (protMassFilterList.Count > 0) Then
                MinimumProteinMassToSearch = CSng(protMassFilterList(0))
            End If

            If (protMassFilterList.Count > 1) Then
                MaximumProteinMassToSearch = CSng(protMassFilterList(1))
            End If

            NumberOfDetectedPeaksToMatch = CInt(.GetParam("match_peak_count"))
            NumberOfAllowedDetectedPeakErrors = CInt(.GetParam("match_peak_allowed_error"))
            MatchedPeakMassTolerance = CSng(.GetParam("match_peak_tolerance"))
            AminoAcidsAllUpperCase = True
            SequenceHeaderInfoToFilter = .GetParam("sequence_header_filter")
            DMS_ID = -1
        End With

    End Sub

    <Obsolete("Unused")>
    Private Function GetDescription() As String

        Dim s As String
        Dim desc As String

        Dim fi As FileInfo
        Dim tr As TextReader

        fi = New FileInfo(m_templateFilePath)
        tr = fi.OpenText
        s = tr.ReadLine
        'Find the correct section block)
        Do While s IsNot Nothing
            If InStr(s, ";DMS_Description = ") > 0 Then
                desc = Mid(s, InStr(s, " = ") + 3)
                Return desc
            End If
            s = tr.ReadLine
        Loop

        Return s

    End Function

    Private Function GetTemplateType() As ParamFileTypes
        Dim s As String
        Dim type As ParamFileTypes

        Dim fi As FileInfo
        Dim tr As TextReader

        fi = New FileInfo(m_templateFilePath)
        tr = fi.OpenText
        s = tr.ReadLine

        Do While s IsNot Nothing
            If InStr(s.ToLower, "num_results = ") > 0 Then
                type = ParamFileTypes.BioWorks_31
                Return type
            Else
                type = ParamFileTypes.BioWorks_20
                Return type
            End If
        Loop

    End Function

    Private Function ReturnMassFilterString(
        minMass As Single,
        maxMass As Single) As String

        Return Format(minMass.ToString, "0") & " " & Format(maxMass.ToString, "0")

    End Function

    Private Function GetFilePath(templateFileName As String) As String
        Return templateFileName
    End Function

End Class

File: ParamFileGenerator/Parameters/Params.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;

namespace ParamFileGenerator
{
    public class Params : IBasicParams, IAdvancedParams
    {
        // Ignore Spelling: Xcalibur

        public enum ParamFileTypes
        {
            BioWorks_20 = 0, // Normal BioWorks 2.0 SEQUEST
            BioWorks_30 = 1, // BioWorks 3.0+ TurboSEQUEST
            BioWorks_31 = 2, // BioWorks 3.1 ClusterQuest
            BioWorks_32 = 3 // BioWorks 3.2 ClusterF***
        }

        private const string DEF_DB_NAME = @"C:\Xcalibur\database\nr.fasta";     // Not really used, just a placeholder

        private const string DEF_ENZ_SECTION_NAME = "SEQUEST_ENZYME_INFO";

        // Basic Parameters
        private ParamFileTypes mType;

        // Advanced Parameters
        private string mIonSeriesString;
        private string mProtMassFilterString;

        private RetrieveParams mFullTemplate;
        private string mTemplateFilePath;

        public static Params BaseLineParamSet => MainProcess.BaseLineParamSet;

        public ParamFileTypes FileType => mType;

        public int DMS_ID { get; set; }

        public string FileName { get; set; }

        public int FileTypeIndex { get; set; }

        public string Description { get; set; }

        public string DefaultFASTAPath { get; set; }

        public string DefaultFASTAPath2 { get; set; }

        public int NumberOfResultsToProcess { get; set; }

        public float PeptideMassTolerance { get; set; }

        public bool CreateOutputFiles { get; set; }

        public IonSeries IonSeries { get; set; }

        public DynamicMods DynamicMods { get; set; }

        public TermDynamicMods TermDynamicMods { get; set; }

        public IsoMods IsotopicModificationsList { get; set; }

        [Obsolete("Use IsotopicModificationsList instead.")]
        public IsoMods IsotopicMods => IsotopicModificationsList;

        public int MaximumNumAAPerDynMod { get; set; }

        public int MaximumNumDifferentialPerPeptide { get; set; }

        public int UsePhosphoFragmentation { get; set; }

        public float FragmentIonTolerance { get; set; }

        public int NumberOfOutputLines { get; set; }

        public int NumberOfDescriptionLines { get; set; }

        public bool ShowFragmentIons { get; set; }

        public bool PrintDuplicateReferences { get; set; }

        public EnzymeDetails SelectedEnzymeDetails { get; set; }

        public int SelectedEnzymeIndex { get; set; }

        public int SelectedEnzymeCleavagePosition { get; set; }

        public IAdvancedParams.FrameList SelectedNucReadingFrame { get; set; }

        public int SelectedNucReadingFrameIndex
        {
            get => (int)SelectedNucReadingFrame;
            set => SelectedNucReadingFrame = (IAdvancedParams.FrameList)value;
        }

        public IBasicParams.MassTypeList ParentMassType { get; set; }

        public IBasicParams.MassTypeList FragmentMassType { get; set; }

        public bool RemovePrecursorPeak { get; set; }

        public float IonCutoffPercentage { get; set; }

        public int MaximumNumberMissedCleavages { get; set; }

        public float MinimumProteinMassToSearch { get; set; }

        public float MaximumProteinMassToSearch { get; set; }

        public int NumberOfDetectedPeaksToMatch { get; set; }

        public int NumberOfAllowedDetectedPeakErrors { get; set; }

        public float MatchedPeakMassTolerance { get; set; }

        public bool AminoAcidsAllUpperCase { get; set; }

        public string PartialSequenceToMatch { get; set; }

        public string SequenceHeaderInfoToFilter { get; set; }

        public int PeptideMassUnits { get; set; }

        public int FragmentMassUnits { get; set; }

        public StaticMods StaticModificationsList { get; set; }

        public List<EnzymeDetails> EnzymeList { get; set; }

        public Hashtable LoadedParamNames { get; set; } = new();

        public void AddLoadedParamName(string parameterName, string parameterValue)
        {
            LoadedParamNames ??= new Hashtable();
            if (!LoadedParamNames.ContainsKey(parameterName))
            {
                LoadedParamNames.Add(parameterName, parameterValue);
            }
        }

        public EnzymeDetails RetrieveEnzymeDetails(int enzymeListIndex)
        {
            return EnzymeList[enzymeListIndex];
        }

        /// <summary>
        /// Constructor
        /// </summary>
        public Params()
        {
            IonSeries = new IonSeries();
            EnzymeList = new List<EnzymeDetails>();
            SelectedEnzymeDetails = new EnzymeDetails();
            DynamicMods = new DynamicMods();
            StaticModificationsList = new StaticMods();
            IsotopicModificationsList = new IsoMods();
            TermDynamicMods = new TermDynamicMods("0.0 0.0");

            MaximumNumDifferentialPerPeptide = 3;
            UsePhosphoFragmentation = 0;
            PeptideMassUnits = (int)IAdvancedParams.MassUnitList.amu;
        }

        public string ReturnMassFilter(float minimumMassToFilter, float maximumMassToFilter)
        {
            return ReturnMassFilterString(minimumMassToFilter, maximumMassToFilter);
        }

        public void LoadTemplate(string templateFileName)
        {
            LoadTemplateParams(templateFileName);
        }

        private void LoadTemplateParams(string templateFileName)
        {
            mTemplateFilePath = GetFilePath(templateFileName);
            var getEnzymeList = new GetEnzymeBlockType(mTemplateFilePath, DEF_ENZ_SECTION_NAME);
            EnzymeList = getEnzymeList.EnzymeList;

            mType = GetTemplateType();

            const string sectionName = "SEQUEST";

            mFullTemplate = new RetrieveParams(mTemplateFilePath);
            // Retrieve Basic Parameters
            mFullTemplate.SetSection(sectionName);
            FileName = Path.GetFileName(mTemplateFilePath);
            SelectedEnzymeIndex = SafeCastInt(mFullTemplate.GetParam("enzyme_number"));
            SelectedEnzymeDetails = EnzymeList[SelectedEnzymeIndex]; // TODO: Add an item first!!
            SelectedEnzymeCleavagePosition = 1;
            MaximumNumberMissedCleavages = SafeCastInt(mFullTemplate.GetParam("max_num_internal_cleavage_sites"));
            ParentMassType = (IBasicParams.MassTypeList)SafeCastInt(mFullTemplate.GetParam("mass_type_parent"));
            FragmentMassType = (IBasicParams.MassTypeList)SafeCastInt(mFullTemplate.GetParam("mass_type_fragment"));
            PartialSequenceToMatch = mFullTemplate.GetParam("partial_sequence");
            DynamicMods = new DynamicMods(mFullTemplate.GetParam("diff_search_options"));
            TermDynamicMods = new TermDynamicMods(mFullTemplate.GetParam("term_diff_search_options"));

            StaticModificationsList = new StaticMods
            {
                // Get Static Mods
                CtermPeptide = SafeCastDouble(mFullTemplate.GetParam("add_Cterm_peptide")),
                CtermProtein = SafeCastDouble(mFullTemplate.GetParam("add_Cterm_protein")),
                NtermPeptide = SafeCastDouble(mFullTemplate.GetParam("add_Nterm_peptide")),
                NtermProtein = SafeCastDouble(mFullTemplate.GetParam("add_Nterm_protein")),
                G_Glycine = SafeCastDouble(mFullTemplate.GetParam("add_G_Glycine")),
                A_Alanine = SafeCastDouble(mFullTemplate.GetParam("add_A_Alanine")),
                S_Serine = SafeCastDouble(mFullTemplate.GetParam("add_S_Serine")),
                P_Proline = SafeCastDouble(mFullTemplate.GetParam("add_P_Proline")),
                V_Valine = SafeCastDouble(mFullTemplate.GetParam("add_V_Valine")),
                T_Threonine = SafeCastDouble(mFullTemplate.GetParam("add_T_Threonine")),
                C_Cysteine = SafeCastDouble(mFullTemplate.GetParam("add_C_Cysteine")),
                L_Leucine = SafeCastDouble(mFullTemplate.GetParam("add_L_Leucine")),
                I_Isoleucine = SafeCastDouble(mFullTemplate.GetParam("add_I_Isoleucine")),
                X_LorI = SafeCastDouble(mFullTemplate.GetParam("add_X_LorI")),
                N_Asparagine = SafeCastDouble(mFullTemplate.GetParam("add_N_Asparagine")),
                O_Ornithine = SafeCastDouble(mFullTemplate.GetParam("add_O_Ornithine")),
                B_avg_NandD = SafeCastDouble(mFullTemplate.GetParam("add_B_avg_NandD")),
                D_Aspartic_Acid = SafeCastDouble(mFullTemplate.GetParam("add_D_Aspartic_Acid")),
                Q_Glutamine = SafeCastDouble(mFullTemplate.GetParam("add_Q_Glutamine")),
                K_Lysine = SafeCastDouble(mFullTemplate.GetParam("add_K_Lysine")),
                Z_avg_QandE = SafeCastDouble(mFullTemplate.GetParam("add_Z_avg_QandE")),
                E_Glutamic_Acid = SafeCastDouble(mFullTemplate.GetParam("add_E_Glutamic_Acid")),
                M_Methionine = SafeCastDouble(mFullTemplate.GetParam("add_M_Methionine")),
                H_Histidine = SafeCastDouble(mFullTemplate.GetParam("add_H_Histidine")),
                F_Phenylalanine = SafeCastDouble(mFullTemplate.GetParam("add_F_Phenylalanine")),
                R_Arginine = SafeCastDouble(mFullTemplate.GetParam("add_R_Arginine")),
                Y_Tyrosine = SafeCastDouble(mFullTemplate.GetParam("add_Y_Tyrosine")),
                W_Tryptophan = SafeCastDouble(mFullTemplate.GetParam("add_W_Tryptophan"))
            };

            // add code to check for existence of isotopic mods

            // Retrieve Advanced Parameters
            mFullTemplate.SetSection(sectionName);
            if (mType == ParamFileTypes.BioWorks_20)
            {
                DefaultFASTAPath = mFullTemplate.GetParam("database_name");
                DefaultFASTAPath2 = "";
                NumberOfResultsToProcess = 500;
            }
            else if (mType == ParamFileTypes.BioWorks_30)
            {
                DefaultFASTAPath = mFullTemplate.GetParam("first_database_name");
                DefaultFASTAPath2 = mFullTemplate.GetParam("second_database_name");
                NumberOfResultsToProcess = SafeCastInt(mFullTemplate.GetParam("num_results"));
            }
            PeptideMassTolerance = (float) SafeCastDouble(mFullTemplate.GetParam("peptide_mass_tolerance"));
            if (mFullTemplate.GetParam("create_output_files") is not null)
            {
                CreateOutputFiles = SafeCastBool(mFullTemplate.GetParam("create_output_files"));
            }
            else
            {
                CreateOutputFiles = true;
            }
            mIonSeriesString = mFullTemplate.GetParam("ion_series");
            IonSeries = new IonSeries(mIonSeriesString);
            MaximumNumAAPerDynMod = SafeCastInt(mFullTemplate.GetParam("max_num_differential_AA_per_mod"));
            if (mType == ParamFileTypes.BioWorks_32)
            {
                MaximumNumDifferentialPerPeptide = SafeCastInt(mFullTemplate.GetParam("max_num_differential_per_peptide"));
            }
            FragmentIonTolerance = (float) SafeCastDouble(mFullTemplate.GetParam("fragment_ion_tolerance"));
            NumberOfOutputLines = SafeCastInt(mFullTemplate.GetParam("num_output_lines"));
            NumberOfDescriptionLines = SafeCastInt(mFullTemplate.GetParam("num_description_lines"));
            ShowFragmentIons = SafeCastBool(mFullTemplate.GetParam("show_fragment_ions"));
            PrintDuplicateReferences = SafeCastBool(mFullTemplate.GetParam("print_duplicate_references"));
            SelectedNucReadingFrame = (IAdvancedParams.FrameList)SafeCastInt(mFullTemplate.GetParam("enzyme_number"));
            RemovePrecursorPeak = SafeCastBool(mFullTemplate.GetParam("remove_precursor_peak"));
            IonCutoffPercentage = (float) SafeCastDouble(mFullTemplate.GetParam("ion_cutoff_percentage"));
            mProtMassFilterString = mFullTemplate.GetParam("protein_mass_filter");

            var protMassFilterList = mProtMassFilterString.Split(' ');
            if (protMassFilterList.Length > 0)
            {
                MinimumProteinMassToSearch = (float) SafeCastDouble(protMassFilterList[0]);
            }

            if (protMassFilterList.Length > 1)
            {
                MaximumProteinMassToSearch = (float) SafeCastDouble(protMassFilterList[1]);
            }

            NumberOfDetectedPeaksToMatch = SafeCastInt(mFullTemplate.GetParam("match_peak_count"));
            NumberOfAllowedDetectedPeakErrors = SafeCastInt(mFullTemplate.GetParam("match_peak_allowed_error"));
            MatchedPeakMassTolerance = (float) SafeCastDouble(mFullTemplate.GetParam("match_peak_tolerance"));
            AminoAcidsAllUpperCase = true;
            SequenceHeaderInfoToFilter = mFullTemplate.GetParam("sequence_header_filter");
            DMS_ID = -1;
        }

        [Obsolete("Unused")]
        private string GetDescription()
        {
            var fi = new FileInfo(mTemplateFilePath);
            TextReader tr = fi.OpenText();
            var s = tr.ReadLine();
            // Find the correct section block)
            while (s is not null)
            {
                if (s.Contains(";DMS_Description = "))
                {
                    return s.Substring(s.IndexOf(" = ", StringComparison.Ordinal) + 3);
                }
                s = tr.ReadLine();
            }

            return "";
        }

        private ParamFileTypes GetTemplateType()
        {
            var fi = new FileInfo(mTemplateFilePath);
            TextReader tr = fi.OpenText();
            var s = tr.ReadLine();

            while (s is not null)
            {
                ParamFileTypes type;
                if (s.ToLower().Contains("num_results = "))
                {
                    type = ParamFileTypes.BioWorks_31;
                    return type;
                }

                type = ParamFileTypes.BioWorks_20;
                return type;
            }

            return default;
        }

        private string ReturnMassFilterString(float minMass, float maxMass)
        {
            return minMass.ToString("0") + " " + maxMass.ToString("0");
        }

        private string GetFilePath(string templateFileName)
        {
            return templateFileName;
        }

        public static int SafeCastInt(object value)
        {
            if (value is int valueInt) return valueInt;
            if (value == null) return 0;

            var valueString = value.ToString().Trim();
            if (string.IsNullOrEmpty(valueString)) return 0;

            if (int.TryParse(valueString, out valueInt)) return valueInt;

            if (double.TryParse(valueString, out var valueDbl))
            {
                return (int) Math.Round(valueDbl);
            }

            return 0;
        }

        public static double SafeCastDouble(string value)
        {
            if (value == null) return 0;

            var valueString = value.Trim();
            if (string.IsNullOrEmpty(valueString)) return 0;

            if (double.TryParse(valueString, out var valueDbl))
            {
                return valueDbl;
            }

            return 0;
        }

        public static bool SafeCastBool(string value)
        {
            if (value == null) return false;

            var valueString = value.Trim();
            if (string.IsNullOrEmpty(valueString)) return false;

            if (bool.TryParse(valueString, out var valueBool))
            {
                return valueBool;
            }

            return false;
        }
    }
}
